README.md
# React Native Numpad

A simple React Native number pad for quickly updating multiple number inputs.

[![npm version](https://badge.fury.io/js/react-native-numpad.svg)](https://badge.fury.io/js/react-native-numpad)
[![runs with expo](https://img.shields.io/badge/Runs%20with%20Expo-000.svg?style=flat-square&logo=EXPO&labelColor=f3f3f3&logoColor=000)](https://expo.io/)

- ✅ **No Dependencies**
- ✅ iOS
- ✅ Android
- ✅ React Native Web
- ✅ JS-Only (No Native Code / No Linking Necessary)

![Screen Recording](screen-recording.gif)

## Demo 👉 <a href="https://snack.expo.io/@git/github.com/glancemoney/react-native-numpad" target="_blank">Expo Snack</a>

## Install

```
yarn add react-native-numpad
```

## Use Cases

- Splitting expenses
- Forms with multiple number inputs
- Spreadsheets
- Calculators

## Usage

```js
import React from 'react';
import NumberPad, { Input, Display } from './index';

export default () => (
  <NumberPad>
    <Display key={0} cursor value={1.0} />
    <Display key={1} cursor value={2.5} />
    <Input />
  </NumberPad>
);
```

## Custom Icons

```js
import React from 'react';
import NumberPad, { Input, Display } from './index';
import { Ionicons } from '@expo/vector-icons';

export default () => (
  <NumberPad>
    <Display key={0} cursor value={1.0} />
    <Display key={1} cursor value={2.5} />
    <Input
      backspaceIcon={<Ionicons icon="ios-backspace" {...Input.iconStyle} />}
      hideIcon={<Ionicons icon="ios-arrow-down" {...Input.iconStyle} />}
    />
  </NumberPad>
);
```

## API

Under the hood, `react-native-numpad` uses the [React Context API](https://reactjs.org/docs/context.html) to link the number inputs (the `<Display>`s) to the number pad (the `<Input>`).

### `<NumberPad>` Component

The `<NumberPad>` component is a [HOC (Higher Order Component)](https://reactjs.org/docs/higher-order-components.html) that does not accept any props besides `children`. It creates a `reactNativeNumpad` context that listens for press events on the number inputs, opens the number input when it detects a press, and then updates the input values when the user presses on the number buttons in the number pad.

### `<Display>` Component

The `<Display>` is the number pad's equivalent of React Native's [`<TextInput>`](https://reactnative.dev/docs/textinput) component. It is a [controlled component](https://reactjs.org/docs/forms.html#controlled-components) that, when pressed, opens the number pad.

| Prop                       | Description                                                                                      | Default |
| -------------------------- | ------------------------------------------------------------------------------------------------ | ------- |
| **`value`**                | Current value of the input (number only)                                                         | _None_  |
| **`style`**                | Any valid style object for [`<TouchableOpacity>`](https://reactnative.dev/docs/touchableopacity) | _None_  |
| **`textStyle`**            | Any valid style object for a [`<Text>`](https://reactnative.dev/docs/text) component             | _None_  |
| **`activeStyle`**          | Any valid style object for a [`<Text>`](https://reactnative.dev/docs/text) component             | _None_  |
| **`invalidTextStyle`**     | Any valid style object for a [`<Text>`](https://reactnative.dev/docs/text) component             | _None_  |
| **`placeholderTextStyle`** | Any valid style object for a [`<Text>`](https://reactnative.dev/docs/text) component             | _None_  |
| **`cursorStyle`**          | Any valid style object for a [`<View>`](https://reactnative.dev/docs/view) component             | _None_  |
| **`blinkOnStyle`**         | Any valid style object for a [`<View>`](https://reactnative.dev/docs/view) component             | _None_  |
| **`blinkOffStyle`**        | Any valid style object for a [`<View>`](https://reactnative.dev/docs/view) component             | _None_  |
| **`onChange`**             | An event handler function that receives the new value (number) as an argument                    | _None_  |
| **`cursor`**               | Whether or not to show the cursor when the input is focused (boolean)                            | true    |
| **`autofocus`**            | Whether or not to autofocus the input when the component is loaded (boolean)                     | false   |

### `<Input>` Component

The `<Input>` a custom number pad keyboard that, unlike the native keyboard, does not minimize when the user presses on a new number input if it is already open. It is stylable and easy to customize.

| Prop                | Description                                                                                                      | Default                  |
| ------------------- | ---------------------------------------------------------------------------------------------------------------- | ------------------------ |
| **`height`**        | Height of the number pad                                                                                         | 270                      |
| **`position`**      | How the number pad will be positioned                                                                            | 'absolute' \| 'relative' |
| **`style`**         | Any valid style object for a [`<View>`](https://reactnative.dev/docs/view) component (`Animated.View`, actually) | _None_                   |
| **`backspaceIcon`** | An Icon element (eg from `react-native-vector-icons` or `@expo/vector-icons`)                                    | _None_                   |
| **`hideIcon`**      | An Icon element (eg from `react-native-vector-icons` or `@expo/vector-icons`)                                    | _None_                   |
| **`onWillHide`**    | Called just before the number pad will hide                                                                      | _None_                   |
| **`onDidHide`**     | Called just after the number pad hides                                                                           | _None_                   |
| **`onWillShow`**    | Called just before the number pad will show                                                                      | _None_                   |
| **`onDidShow`**     | Called just after the number pad shows                                                                           | _None_                   |

### `<AvoidingView>` Component

Sometimes React Native's built-in [<KeyboardAvoidingView>](https://reactnative.dev/docs/keyboardavoidingview) does not work smoothly with the number pad: it can either have performance issues where animations are choppy or it can be difficult to configure its height properly altogether. We've included a number pad context-aware version that adjusts it's height based on the keyboard animation to achieve a smooth frame rate.

| Prop        | Description                                                                                                      | Default |
| ----------- | ---------------------------------------------------------------------------------------------------------------- | ------- |
| **`style`** | Any valid style object for a [`<View>`](https://reactnative.dev/docs/view) component (`Animated.View`, actually) | _None_  |

## Version History (Change Log)

View [here](./CHANGELOG.md).

## Contribute

We welcome contributions! If you are interested in contributing, consider helping us with one of the following tasks:

- Rewrite components in TypeScript using arrow-function components and [React hooks](https://reactjs.org/docs/hooks-intro.html)
- Add TypeScript bindings
- Add Tests

## Glance Money

[![Glance Money Logo](https://uploads-ssl.webflow.com/5ec80cb02fe1c031a342c6cc/5ecdb5f349bce13a545a2dae_Artboard%20Copy%2062.png)](https://glance.money)

We wrote this for, actively use, and maintain this library for [Glance Money](https://glance.money). Now it is free and open for the world to use ❤️

## License

[MIT licensed.](./LICENSE)


CHANGELOG.md
# Change Log

- 0.3.0 - Migrate to TypeScript + publish TypeScript definitions
- 0.2.0 - Initial published version


lib/Display.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, View, Text } from 'react-native';

import NumberPadContext from './NumberPadContext';
import styles from './styles';

const parse = (string) => {
  return parseFloat(string.replace(/,/g, ''));
};

const format = (string, initial) => {
  let decimal = string.includes('.');
  let [whole = '', part = ''] = string.split('.');
  decimal = initial && string !== '0' ? true : decimal;
  whole = whole.replace(/,/g, '').substring(0, 9);
  whole = whole ? parseInt(whole).toLocaleString('en-US') : '0';
  part = part.substring(0, 2);
  part = initial && decimal ? part.padEnd(2, '0') : part;
  return `${whole}${decimal ? '.' : ''}${part}`;
};

export default class Display extends Component {
  static contextType = NumberPadContext;

  static propTypes = {
    value: PropTypes.number.isRequired,
    style: PropTypes.object.isRequired,
    textStyle: PropTypes.object.isRequired,
    activeStyle: PropTypes.object.isRequired,
    activeTextStyle: PropTypes.object.isRequired,
    invalidTextStyle: PropTypes.object.isRequired,
    placeholderTextStyle: PropTypes.object.isRequired,
    cursorStyle: PropTypes.object.isRequired,
    blinkOnStyle: PropTypes.object.isRequired,
    blinkOffStyle: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    isValid: PropTypes.func.isRequired,
    cursor: PropTypes.bool.isRequired,
    autofocus: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    value: 0.0,
    style: styles.display,
    textStyle: styles.displayText,
    activeStyle: styles.activeDisplay,
    activeTextStyle: styles.activeDisplayText,
    invalidTextStyle: styles.invalidDisplayText,
    placeholderTextStyle: styles.placeholderDisplayText,
    cursorStyle: styles.cursor,
    blinkOnStyle: styles.blinkOn,
    blinkOffStyle: styles.blinkOff,
    onChange: () => {},
    isValid: () => true,
    cursor: false,
    autofocus: false,
  };

  constructor(props) {
    super(props);

    this.blink = null;

    const value = format(String(this.props.value), true);

    this.state = {
      valid: true,
      active: false,
      blink: true,
      value,
      lastValue: value,
      empty: value === '0',
    };
  }

  componentDidMount() {
    this.context.registerDisplay(this);
    if (this.props.autofocus) {
      setTimeout(this.focus, 0); // setTimeout fixes an issue with it sometimes not focusing
    }
  }

  componentWillUnmount() {
    this.context.unregisterDisplay(this);
    if (this.blink) clearInterval(this.blink);
  }

  focus = (propagate = true) => {
    if (propagate) this.context.focus(this);
    this.setState({
      active: true,
      lastValue: format(this.state.value, true),
      value: '0',
    });
    if (this.props.cursor) {
      if (this.blink) clearInterval(this.blink);
      this.blink = setInterval(() => {
        this.setState({
          blink: !this.state.blink,
        });
      }, 600);
    }
  };

  blur = (propagate = true) => {
    if (propagate && this.context.display === this._reactInternalFiber.key) {
      this.context.blur();
    }

    const value = format(this.state.value, true);

    this.setState({
      active: false,
      value: this.value(value),
    });
  };

  empty = (value) => {
    value = value ? value : this.state.value;
    return value === '0';
  };

  value = (value) => {
    value = value ? value : this.state.value;
    return this.empty(value) ? this.state.lastValue : value;
  };

  onInputEvent = (event) => {
    const value = format(
      event === 'backspace'
        ? this.state.value.substring(0, this.state.value.length - 1)
        : `${this.state.value}${event}`
    );
    const valid = this.props.isValid(value);
    this.setState({
      value,
      valid,
    });
    this.props.onChange(parse(this.value(value)));
  };

  render() {
    const { valid, value, active } = this.state;
    const empty = this.empty();
    const blink = this.state.blink
      ? this.props.blinkOnStyle
      : this.props.blinkOffStyle;
    const style = [
      { flexDirection: 'row' },
      this.props.style,
      active ? this.props.activeStyle : null,
    ];
    const textStyle = [
      this.props.textStyle,
      active ? this.props.activeTextStyle : null,
    ];
    const cursorStyle = [this.props.cursorStyle];
    return (
      <TouchableOpacity style={style} onPress={this.focus}>
        <View style={[cursorStyle, active ? blink : null]}>
          <Text
            style={[
              textStyle,
              valid ? null : this.props.invalidTextStyle,
              empty ? this.props.placeholderTextStyle : null,
            ]}
          >
            {empty ? this.state.lastValue : value}
          </Text>
        </View>
      </TouchableOpacity>
    );
  }
}


lib/AvoidingView.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import NumberPadContext from './NumberPadContext';

export default class AvoidingView extends Component {
  static contextType = NumberPadContext;

  static propTypes = {
    children: PropTypes.any,
    style: PropTypes.object,
  };

  constructor(props) {
    super(props);

    this.animation = new Animated.Value(0);
  }

  show = () => {
    Animated.timing(this.animation, {
      duration: 200,
      toValue: this.context.height,
      useNativeDriver: false,
    }).start();
  };

  hide = () => {
    Animated.timing(this.animation, {
      duration: 200,
      toValue: 0,
      useNativeDriver: false,
    }).start();
  };

  componentDidMount() {
    this.context.registerAvoidingView(this);
  }

  componentWillUnmount() {
    Animated.timing(this.animation, {
      duration: 200,
      toValue: 0,
      useNativeDriver: false,
    }).start();
  }

  render() {
    return (
      <Animated.View
        style={[
          {
            paddingBottom: this.animation.interpolate({
              inputRange: [0, this.context.height],
              outputRange: [0, this.context.height],
            }),
          },
          this.props.style,
        ]}
      >
        {this.props.children}
      </Animated.View>
    );
  }
}


lib/NumberPadContext.ts
import * as React from 'react';
import type Input from './Input';
import type Display from './Display';
import type AvoidingView from './AvoidingView';

type NumberPadContextType = {
  display: null | string;
  input: null | Input;
  height: number;
  focus: (display: Display) => void;
  blur: () => void;
  onInputEvent: (ev: string) => void;
  registerDisplay: (display: Display) => void;
  unregisterDisplay: (display: Display) => void;
  registerAvoidingView: (view: AvoidingView) => void;
  unregisterAvoidingView: (view: AvoidingView) => void;
  registerInput: (input: Input) => void;
  setHeight: (height: number) => void;
};

const nullFn = () => {};

const defaultContext: NumberPadContextType = {
  display: null,
  input: null,
  height: 0,
  focus: nullFn,
  blur: nullFn,
  onInputEvent: nullFn,
  registerDisplay: nullFn,
  unregisterDisplay: nullFn,
  registerAvoidingView: nullFn,
  unregisterAvoidingView: nullFn,
  registerInput: nullFn,
  setHeight: nullFn,
};

export default React.createContext(defaultContext);


lib/Input.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, View, Text, Animated } from 'react-native';

import NumberPadContext from './NumberPadContext';
import styles from './styles';

const inputs = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0'];

const Icon = (props) => {
  try {
    const { Ionicons } = require('@expo/vector-icons');
    return <Ionicons {...props} />;
  } catch (error) {
    try {
      const Ionicons = require('react-native-vector-icons/Ionicons');
      return <Ionicons {...props} />;
    } catch (error) {
      console.warn(
        'react-native-numer-pad: you need to install @expo/vector-icons or react-native-vector-icons for icons to work.'
      );
    }
  }
};

export default class Input extends Component {
  static contextType = NumberPadContext;

  static propTypes = {
    height: PropTypes.number,
    position: PropTypes.oneOf(['relative', 'absolute']).isRequired,
    style: PropTypes.object,
    onWillHide: PropTypes.func,
    onDidHide: PropTypes.func,
    onWillShow: PropTypes.func,
    onDidShow: PropTypes.func,
  };

  static defaultProps = {
    height: 270,
    position: 'absolute',
  };

  constructor(props) {
    super(props);

    this.animation = new Animated.Value(0);
  }

  show = () => {
    if (this.props.onWillShow) this.props.onWillShow();
    Animated.timing(this.animation, {
      duration: 200,
      toValue: this.props.height,
      useNativeDriver: true,
    }).start(this.props.onDidShow);
  };

  hide = () => {
    if (this.props.onWillHide) this.props.onWillHide();
    Animated.timing(this.animation, {
      duration: 200,
      toValue: 0,
      useNativeDriver: true,
    }).start(this.props.onDidHide);
  };

  componentDidMount() {
    this.context.registerInput(this);
    this.context.setHeight(this.props.height);
  }

  componentWillUnmount() {
    Animated.timing(this.animation, {
      duration: 200,
      toValue: 0,
      useNativeDriver: true,
    }).start();
  }

  getStyle = () => {
    const interpolation = this.animation.interpolate({
      inputRange: [0, this.props.height],
      outputRange: [this.props.height, 0],
    });
    return this.props.position === 'absolute'
      ? {
          position: 'absolute',
          bottom: 0,
          height: this.props.height,
          transform: [
            {
              translateY: interpolation,
            },
          ],
        }
      : {
          height: interpolation,
        };
  };

  render() {
    return (
      <Animated.View style={[this.getStyle(), this.props.style]}>
        <View style={styles.input}>
          <View style={styles.pad}>
            {inputs.map((value, index) => {
              return (
                <TouchableOpacity
                  key={index}
                  style={styles.button}
                  onPress={() => this.context.onInputEvent(value)}
                >
                  <Text style={styles.buttonText}>{value}</Text>
                </TouchableOpacity>
              );
            })}
            <TouchableOpacity
              key="backspace"
              style={styles.button}
              onPress={() => this.context.onInputEvent('backspace')}
            >
              <Icon
                name="ios-backspace"
                size={
                  styles.buttonText.fontSize ||
                  36 /* fixes bug where react-native-web can't access stylesheet properties */
                }
                color={
                  styles.buttonText.color ||
                  '#888' /* fixes bug where react-native-web can't access stylesheet properties */
                }
              />
            </TouchableOpacity>
          </View>
          <TouchableOpacity style={styles.hide} onPress={this.context.blur}>
            <Icon name="ios-arrow-down" size={36} color="#aaa" />
          </TouchableOpacity>
        </View>
      </Animated.View>
    );
  }
}
